### DISCUSSION TITLE
The making of a scrollable, dynamically size drawer

### DISCUSSION THREAD

**Discussion by ororsatti** [gorhom/react-native-bottom-sheet#1363] [Ideas]
Hello @gorhom,
Using your lib I was able to create a scrollable, dynamically size drawer that:
- Can have maximum height. (havent tested minimum but I assume its possible too)
- resizes according to your content size
- is scrollable 

I hit a couple of road bumps on the development,
and i need your help to proceed with this idea.

Is it possible we will chat further on this ? This library helped us alot in our development,
and it seems its a long requested feature. 
(https://github.com/gorhom/react-native-bottom-sheet/issues/658 , https://github.com/gorhom/react-native-bottom-sheet/issues/32)

Im finishing with a visual example of the described above behavior, 

https://github.com/gorhom/react-native-bottom-sheet/assets/109456266/d145d6ea-66ca-46ba-9df1-72583670cfd5

Hope you liked it,
Please contact me so we can make this possible.
Email address: letscodetogether@skiff.com

**Comment by ororsatti** [comment]
@gorhom 
Panning motion is working as well, but your lib is quite complex. 
Would you like to discuss on it ?

**Comment by DipoAreoye** [comment]
Hey can I take a look at your implementation?

**Reply by ororsatti** [reply]
Sure, want me to share it here or something?

**Reply by ororsatti** [reply]
https://github.com/gorhom/react-native-bottom-sheet/pull/1402

**Reply by g-todorov** [reply]
Great work, just tried it with `BottomSheetScrollView` in my code and works like charm.
Did you have any success with `BottomSheetFlatList`? Seems that `contentHeight` is again an issue like most of the workaround solutions so far.

**Reply by ororsatti** [reply]
Hello, thanks for the kind words @g-todorov .
I have tested it with flat list and it seems to work just fine, heres a video:

https://github.com/gorhom/react-native-bottom-sheet/assets/109456266/d86e404e-fbc4-4319-b645-b029edd6a01e

I have even added a sticky header for good measure.

Make sure you pass the `innerScrollViewAnimatedStyles` from the hook to the list's styles and not override the `maxHeight` style since the hook relay on it. I might need to change the name since its a little bit misleading.

![Screenshot 2023-06-17 at 1 21 43](https://github.com/gorhom/react-native-bottom-sheet/assets/109456266/e08b310f-ec8d-46c4-8d19-a4e9f5a5cc0f)

Any review/more testing will be appreciated. 
Thanks.

**Reply by g-todorov** [reply]
You are right, it works as expected. Probably it was a misconfiguration on my end. Great work again and I hope it gets merged soon.

**Comment by Eli-Nathan** [comment]
I found some problems with the solution in the linked PR when using this hook with dynamic content that changes before showing the bottom sheet.

The solution I opted for is:

```ts
// ....
const initialSnapPoints = useMemo(() => ["CONTENT_HEIGHT"], [])
const {
    animatedHandleHeight,
    animatedSnapPoints,
    animatedContentHeight,
    handleContentLayout,
} = useBottomSheetDynamicSnapPoints(initialSnapPoints)

const maxHeightStyle = useBottomSheetMaxHeight(maxHeight)

return (
    <BottomSheetModal
        // .... other props
        snapPoints={animatedSnapPoints as any}
        handleHeight={animatedHandleHeight}
        contentHeight={animatedContentHeight}
    >
        <BottomSheetScrollView
            onLayout={handleContentLayout}
            style={{
                maxHeight: maxHeightStyle,
            }}
        >
            {children}
        </BottomSheetScrollView>
    </BottomSheetModal>
)
// ...
```

With the custom `useBottomSheetMaxHeight` hook looking like this:

```ts
/*
    Designed to take the specified max height, take away top safe
    area inset and convert to a decimal
*/
import { useWindowDimensions } from "react-native"
import { useSafeAreaInsets } from "react-native-safe-area-context"

export const useBottomSheetMaxHeight = (maxHeight = "85%"): number => {
    const { height: deviceHeight } = useWindowDimensions()
    const maxHeightDecimal = parseFloat(`${maxHeight}`) / 100.0
    const { top: topSafeAreaInset } = useSafeAreaInsets()

    return (deviceHeight - topSafeAreaInset) * maxHeightDecimal
}

```

Hope that helps someone 😄

**Reply by ororsatti** [reply]
hmm looks actually awesome.
how does it behave with lists?

**Reply by Eli-Nathan** [reply]
Tried with library's scrollview component but not `<FlatList />`
It also doesn't have the ability like your solution to provide a number rather than a percentage.
Although that should be a small condition to add by checking the `typeof maxHeight`

**Reply by ororsatti** [reply]
I wonder if itll be more effective to simply migrate this PR to be your code suggestion.

**Reply by Eli-Nathan** [reply]
Sure feel free!
Although I'm not 100% sure if it's a hook this small then it'd be worth adding to the lib or not. It's more about specific use cases. This relies on a bottom sheet to cover the whole app including bottom tabs.
Some use cases for bottom sheet would place them within certain screens where the device height check wouldn't be sufficient.
What do you think?

**Reply by Eli-Nathan** [reply]
Or even if the `useBottomSheetDynamicSnapPoints` hook could take an optional `maxHeight` param and return the maxHeight style from there? 🤔

**Reply by Eli-Nathan** [reply]
I think both my and your original (awesome btw) solution only _really_ work when there's just 1 snapPoint of "CONTENT_HEIGHT"
If you add one that's higher than the computed content height it doesn't behave so great

**Reply by Eli-Nathan** [reply]
For example: if the computed max height makes the sheet span (roughly) 90% of the window height (minus safe areas) and the snapPoints are set to `["CONTENT_HEIGHT", "99%"]` then you get the following behaviour:
Maybe that just relies on developers using sensible values though?

![Gif bottom sheet](https://github.com/gorhom/react-native-bottom-sheet/assets/32842898/d27ba14b-0211-4483-8cc6-8d48c0513e01)

**Reply by ororsatti** [reply]
intresting point,
quite overwhelmed currently by work, but id like to take a look into this issue further

**Reply by Eli-Nathan** [reply]
Were you able to run the example app when forking to the repo? I've got a solution with extending the`useBottomSheetDynamicSnapPoints` hook but can't get any of the examples running. Not sure if the contributing guide is out of date.

**Reply by Eli-Nathan** [reply]
https://github.com/gorhom/react-native-bottom-sheet/pull/1510 for reference.
It just pulls most of the logic from my example above (extended to allow a number value), calls it from the `useBottomSheetDynamicSnapPoints` hook and returns a child view style with max height that can be applied to the View/ScrollView/FlatList

**Comment by Eli-Nathan** [comment]
https://github.com/gorhom/react-native-bottom-sheet/pull/1510 Raised this PR to allow both View's and Scrollables to work with dynamic heights and to apply an optional maximum height.

Hard to tell if @gorhom is still active on this library or not but if so please could you take a look?

**Comment by gorhom** [comment]
Sorry I've been busy with my daily job, I'll have a look at it tomorrow 👍

**Reply by Eli-Nathan** [reply]
Perfect thanks very much!!
